Tools/Algorithms/psatd.ipynb

```diff
@@ -0,0 +1,631 @@
+    "t = sp.symbols(r't', real=True, positive=True)\n",
+    "tn = sp.symbols(r't_n', real=True, positive=True)\n",
+    "dt = sp.symbols(r'\\Delta{t}', real=True, positive=True)\n",
+    "\n",
+    "# The assumption that kx, ky and kz are positive is general enough\n",
+    "# and makes it easier for SymPy to perform some of the calculations\n",
+    "kx = sp.symbols(r'k_x', real=True, positive=True)\n",
+    "ky = sp.symbols(r'k_y', real=True, positive=True)\n",
+    "kz = sp.symbols(r'k_z', real=True, positive=True)\n",
+    "\n",
+    "# Define symbols for the Cartesian components of the electric field\n",
+    "Ex = sp.symbols(r'E^x')\n",
+    "Ey = sp.symbols(r'E^y')\n",
+    "Ez = sp.symbols(r'E^z')\n",
+    "E = Matrix([[Ex], [Ey], [Ez]])\n",
+    "\n",
+    "# Define symbols for the Cartesian components of the magnetic field\n",
+    "Bx = sp.symbols(r'B^x')\n",
+    "By = sp.symbols(r'B^y')\n",
+    "Bz = sp.symbols(r'B^z')\n",
+    "B = Matrix([[Bx], [By], [Bz]])\n",
+    "\n",
+    "# Define symbol for the scalar field F used with div(E) cleaning\n",
+    "if divE_cleaning:\n",
+    "    F = sp.symbols(r'F')\n",
+    "\n",
+    "# Define symbol for the scalar field G used with div(B) cleaning\n",
+    "if divB_cleaning:\n",
+    "    G = sp.symbols(r'G')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### First-order ODEs for $\\boldsymbol{E}$,  $\\boldsymbol{B}$, $F$ and $G$:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Define first-order time derivatives of the electric field\n",
+    "dEx_dt = I*c**2*(ky*Bz-kz*By) \n",
+    "dEy_dt = I*c**2*(kz*Bx-kx*Bz)\n",
+    "dEz_dt = I*c**2*(kx*By-ky*Bx)\n",
+    "\n",
+    "# Define first-order time derivatives of the magnetic field\n",
+    "dBx_dt = -I*(ky*Ez-kz*Ey)\n",
+    "dBy_dt = -I*(kz*Ex-kx*Ez)\n",
+    "dBz_dt = -I*(kx*Ey-ky*Ex)\n",
+    "\n",
+    "# Define first-order time derivative of the scalar field F used with div(E) cleaning,\n",
+    "# and related additional terms in the first-order time derivative of the electric field\n",
+    "if divE_cleaning:\n",
+    "    dEx_dt += I*c**2*F*kx \n",
+    "    dEy_dt += I*c**2*F*ky\n",
+    "    dEz_dt += I*c**2*F*kz\n",
+    "    dF_dt = I*(kx*Ex+ky*Ey+kz*Ez)\n",
+    "\n",
+    "# Define first-order time derivative of the scalar field G used with div(B) cleaning,\n",
+    "# and related additional terms in the first-order time derivative of the magnetic field\n",
+    "if divB_cleaning:\n",
+    "    dBx_dt += I*c**2*G*kx\n",
+    "    dBy_dt += I*c**2*G*ky\n",
+    "    dBz_dt += I*c**2*G*kz\n",
+    "    dG_dt = I*(kx*Bx+ky*By+kz*Bz)\n",
+    "\n",
+    "# Define array of first-order time derivatives of the electric and magnetic fields\n",
+    "dE_dt = Matrix([[dEx_dt], [dEy_dt], [dEz_dt]])\n",
+    "dB_dt = Matrix([[dBx_dt], [dBy_dt], [dBz_dt]])"
```

Comment:
It seems that these variables are not used and could be removed.

Reply:
It is true that these variables are not used in the notebook. However, I think we can expect this notebook to be used in an interactive way (namely, by running individual cells and checking results) and I think it is useful/handy to have access to the first time derivatives of E and B independently through `dE_dt` and `dB_dt`, rather than having access to them only through the more general array `dEBFG_dt`. This would be my rationale for keeping these variables available to the user.